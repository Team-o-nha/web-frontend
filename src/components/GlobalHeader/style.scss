@import "@/styles/custom.scss";
.menuBar {
  position: relative;
  padding: 0 20px;
  border-bottom: solid 1px #e8e8e8;
  overflow: auto;
  box-shadow: 0 0 30px #f3f1f1;
}
.logo {
  width: 150px;
  float: left;
  img {
    width: 100%;
    display: inline-block;
    padding: 15px 20px;
    text-transform: capitalize;
  }
}
.menuCon {
  width: calc(100% - 150px);
  float: left;
}
.menuCon .ant-menu-item {
  padding: 0px 5px;
}
.menuCon .ant-menu-submenu-title {
  padding: 10px 20px;
}
.menuCon .ant-menu-item a,
.menuCon .ant-menu-submenu-title a {
  padding: 10px 15px;
}
.menuCon .ant-menu-horizontal {
  border-bottom: none;
}
.menuCon .leftMenu {
  float: left;
}
.menuCon .rightMenu {
  float: right;
}
.barsMenu {
  float: right;
  height: 32px !important;
  padding: 6px !important;
  margin-top: 8px;
  display: none !important;
  background: none !important;
}
.barsBtn {
  display: block !important;
  width: 20px !important;
  height: 2px !important;
  background: #607d8b !important;
  position: relative !important;
}
.barsBtn:after,
.barsBtn:before {
  content: attr(x) !important;
  width: 20px !important;
  position: absolute !important;
  top: -6px !important;
  left: 0 !important;
  height: 2px;
  background: #607d8b !important;
}
.barsBtn:after {
  top: auto !important;
  bottom: -6px !important;
}
.ant-drawer-body {
  padding: 0;
}
.ant-drawer-body .ant-menu-horizontal > .ant-menu-item,
.ant-drawer-body .ant-menu-horizontal > .ant-menu-submenu {
  display: inline-block;
  width: 100%;
}
.ant-menu-horizontal > .ant-menu-item-selected,
.ant-menu-horizontal > .ant-menu-submenu-selected {
  > a {
    color: #607d8b !important;
  }
  color: #607d8b !important;
  border-bottom: 2px solid #607d8b !important;
}

.ant-drawer-body .ant-menu-horizontal {
  border-bottom: none;
}
.ant-drawer-body .ant-menu-horizontal > .ant-menu-item:hover {
  border-bottom-color: transparent;
}

@media (max-width: 767px) {
  .barsMenu {
    display: inline-block !important;
  }
  .leftMenu,
  .rightMenu {
    display: none;
  }
  .logo a {
    margin-left: -20px;
  }
  .menuCon .ant-menu-item,
  .menuCon .ant-menu-submenu-title {
    padding: 1px 20px;
  }
  .logo a {
    padding: 10px 20px;
  }
}
