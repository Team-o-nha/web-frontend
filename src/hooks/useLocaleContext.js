// libs import
import { useContext } from "react";
import LocaleContext from "@/contexts/LocaleContext";

// ご参考に：https://qiita.com/Tokky0425/items/70746f57f4affba0660c
// Why export connt (not export default)?
// Cuz this is the only way to test useContext
// Which requires a custom hook to wrapper over useContext
export const useLocaleContext = () => useContext(LocaleContext);
