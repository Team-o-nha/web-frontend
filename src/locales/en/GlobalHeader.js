export default {
  "header.home": "Home",
  "header.reservation": "Reservation",
  "header.vacancy": "Vacancy Inquiry"
};
