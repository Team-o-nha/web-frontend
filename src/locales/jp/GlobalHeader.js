export default {
  "header.home": "ホーム",
  "header.reservation": "予約",
  "header.vacancy": "空室問い合わせ"
};
