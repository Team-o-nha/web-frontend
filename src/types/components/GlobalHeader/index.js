// @flow

/**
 * GlobalHeader 型定義
 * pathname: current location pathname
 */
export type GLOBAL_HEADER_TYPE = {
  pathname: string
};
