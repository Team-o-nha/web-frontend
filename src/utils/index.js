// libs import
import React, { createElement } from "react";
import { Row, Col, Form, Typography, message } from "antd";
// others
import cookieUtils from "./cookieUtils";

// generate components
const getComponentsToRender = ({ components, componentProps }) =>
  components.map(child =>
    createElement(child.component, { key: child.key, ...componentProps }, null)
  );

// optimize generate form
const getFormDataToRender = ({ formWrapper }) => (
  <div {...formWrapper}>
    <Row>
      {formWrapper.children.formItems.map(item => (
        <Col key={item.name} {...item.grid}>
          <Form.Item
            className={item.className}
            label={<Typography.Text>{item.label.children}</Typography.Text>}
            colon={false}
          >
            {item.content}
          </Form.Item>
        </Col>
      ))}
    </Row>
  </div>
);

// optimize generate form advanced
const getFormDataToRenderAdvanced = ({ formWrapper }) => {
  const ItemGenerator = item => (
    <Form.Item
      className={item.className}
      label={
        item.label ? (
          <Typography.Text>{item.label.children}</Typography.Text>
        ) : (
          ""
        )
      }
      colon={false}
    >
      {item.content}
    </Form.Item>
  );
  const ColGenerator = item => (
    <Col key={item.name} {...item.grid}>
      {ItemGenerator(item)}
    </Col>
  );
  return (
    <div {...formWrapper}>
      <Row>
        {formWrapper.children.formItems.map(item => (
          <Col key={item.key} {...item.grid}>
            {item.children
              ? item.children.map(child => ColGenerator(child))
              : ItemGenerator(item)}
          </Col>
        ))}
      </Row>
    </div>
  );
};

// generate DOW
const generateDOW = index => {
  switch (index) {
    case 1:
      return "月";
    case 2:
      return "火";
    case 3:
      return "水";
    case 4:
      return "木";
    case 5:
      return "金";
    case 6:
      return "土";
    case 0:
      return "日";
    default:
  }
  return "月";
};

// is saturday or sunday
const isSatOrSun = dowLabel => ({
  // 6: saturday
  isSatur: dowLabel === 6,
  // 0: sunday
  isSun: dowLabel === 0
});

// default http response callback error
const defaultHttpResponseCbError = (url, err) => {
  message.error("失敗。LOGでご確認お願い致します。");
  // eslint-disable-next-line no-console
  console.log("エラーが発生したURL：", url);
  // eslint-disable-next-line no-console
  console.log("エラー情報：", err);
};

const generateFunctionButtonKeys = (componentName, prefixes) => {
  let componentObj = {};
  prefixes.forEach(prefix => {
    let keyObj = {};
    for (let i = 1; i <= 12; i += 1) {
      keyObj = {
        ...keyObj,
        [`${prefix}_KEY_F${i}`]: `${componentName}_${prefix}_KEY_F${i}`
      };
    }
    componentObj = { ...componentObj, ...keyObj };
  });
  return {
    [componentName]: { ...componentObj }
  };
};

// find rendered component on DOM by data-test attribute
const findByTestAttr = (component, attr) => {
  const wrapper = component.find(`[data-test='${attr}']`);
  return wrapper;
};

/**
 * Define config options
 * @typedef {Object} formatterConfig
 * @property {number} minimumFractionDigits - the number of decimal digits
 */
/**
 * format numbers. For example:
 * 3000.3234353132 -> "3,000.323"
 * 2.0 -> "2.0", 2.012345 -> "2.012" (using minimumFractionDigits = 0, default maximumFractionDigits = 3)
 * @param {number} number - number needs to be converted to string
 * @param {formatterConfig} options - formatter configs
 */
const formatNumber = (number, formatterConfig = { minimumFractionDigits: 0 }) =>
  new Intl.NumberFormat("en-US", formatterConfig).format(number);

export {
  cookieUtils,
  getComponentsToRender,
  getFormDataToRender,
  getFormDataToRenderAdvanced,
  generateDOW,
  isSatOrSun,
  defaultHttpResponseCbError,
  generateFunctionButtonKeys,
  findByTestAttr,
  formatNumber
};
