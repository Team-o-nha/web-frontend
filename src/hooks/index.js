import { useLocaleContext } from "./useLocaleContext";
import { useEventListener } from "./useEventListener";
import { useRouter } from "./useRouter";
import { useQuery } from "./useQuery";
import { useWhyDidYouUpdate } from "./useWhyDidYouUpdate";

export {
  useLocaleContext,
  useEventListener,
  useRouter,
  useQuery,
  useWhyDidYouUpdate
};
