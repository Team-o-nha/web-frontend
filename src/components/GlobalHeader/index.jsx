// @flow
// libs import
import React, { useState } from "react";
import { Link } from "react-router-dom";
import { Menu, Button, Drawer } from "antd";
// hooks
import { useLocaleContext } from "@/hooks";
// types
import { type GLOBAL_HEADER_TYPE } from "@/types/components/GlobalHeader";
// others
import logo from "../../assets/images/logo.webp";
import CONSTANTS from "@/constants";
import "./style.scss";

/**
 * GlobalHeader props 型定義
 * @param {string} pathname
 */
const GlobalHeader = ({ pathname }: GLOBAL_HEADER_TYPE) => {
  // define urlEndpoint by substring pathname
  // case 1: pathname == ["/en/xxx", "/jp/xxx"]
  // case 2: pathname == ["/xxx"]
  // endpoint = /xxx
  const urlEndpoint = CONSTANTS.LOCALE.PATH_REGEX.test(pathname)
    ? pathname.substring(CONSTANTS.LOCALE.LOCALE_LENGTH + 1, pathname.length)
    : pathname;
  // define selectedMenuKey
  // case 1: endpoint === "/" -> this is Home page
  // case 2: endpoint === "/xxx" -> substring to get xxx and capitalize (Xxx)
  const selectedMenuKey =
    urlEndpoint === "/" || urlEndpoint === ""
      ? CONSTANTS.PAGE_NAME.HOME
      : urlEndpoint
          .substring(1, urlEndpoint.length)
          .toLowerCase()
          .replace(/(^| )(\w)/g, s => s.toUpperCase());

  // get locale, localeDataSource from context
  // FIXME: using hooks from inside LocaleContext, need to move to /hooks folder?
  const { locale, localeDataSource } = useLocaleContext();
  const [visible, setVisible] = useState(false);
  const showDrawer = () => {
    setVisible(true);
  };
  const onClose = () => {
    setVisible(false);
  };
  return (
    <nav className="menuBar">
      <div className="logo">
        <img src={logo} alt="logo" />
      </div>
      <div className="menuCon">
        <div className="leftMenu">
          <Menu mode="horizontal" selectedKeys={[selectedMenuKey]}>
            <Menu.Item key={CONSTANTS.PAGE_NAME.HOME}>
              <Link to={`/${locale}/`}>{localeDataSource["header.home"]}</Link>
            </Menu.Item>
          </Menu>
        </div>
        <div className="rightMenu">
          <Menu mode="horizontal" selectedKeys={[selectedMenuKey]}>
            <Menu.Item key="i18next-jp" className="header-menu-item-right">
              <a href={`/${CONSTANTS.LOCALE.JA_JP}${urlEndpoint}`}>日本語</a>
            </Menu.Item>
            <Menu.Item key="i18next-en" className="header-menu-item-right">
              <a href={`/${CONSTANTS.LOCALE.EN_US}${urlEndpoint}`}>English</a>
            </Menu.Item>
          </Menu>
        </div>
        <Button className="barsMenu" type="primary" onClick={showDrawer}>
          <span className="barsBtn"></span>
        </Button>
        <Drawer
          placement="right"
          closable={false}
          onClose={onClose}
          visible={visible}
        >
          <Menu>
            <Menu.Item key={CONSTANTS.PAGE_NAME.HOME}>
              <Link to={`/${locale}/`}>{localeDataSource["header.home"]}</Link>
            </Menu.Item>
          </Menu>
          <Menu>
            <Menu.Item key="i18next-jp" className="header-menu-item-right">
              <a href={`/${CONSTANTS.LOCALE.JA_JP}${urlEndpoint}`}>日本語</a>
            </Menu.Item>
            <Menu.Item key="i18next-en" className="header-menu-item-right">
              <a href={`/${CONSTANTS.LOCALE.EN_US}${urlEndpoint}`}>English</a>
            </Menu.Item>
          </Menu>
        </Drawer>
      </div>
    </nav>
  );
};

export default GlobalHeader;
