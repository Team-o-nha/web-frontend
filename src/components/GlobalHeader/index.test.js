// @flow
// libs import
import React from "react";
import { shallow } from "enzyme";
// components import
import GlobalHeader from "./index";
// hooks
import * as LocaleContext from "@/hooks/useLocaleContext";
// others
import { findByTestAttr } from "@/utils";
import localeDataSourceJP from "@/locales/jp/GlobalHeader";

const setUp = () => {
  const props = {
    pathname: "/jp/reservation"
  };
  const enzymeWrapper = shallow(<GlobalHeader {...props} />);
  return {
    enzymeWrapper
  };
};

/**
 * GlobalHeader testing
 */
describe("GlobalHeader", () => {
  let globalHeaderEnzymeWrapper;
  /**
   * shallow render GlobalHeader before each test
   */
  beforeEach(() => {
    const { enzymeWrapper } = setUp();
    globalHeaderEnzymeWrapper = enzymeWrapper;
  });

  /**
   * reset mock modules before each test
   */
  beforeEach(() => {
    jest.resetModules();
  });

  /**
   * Shallow render with pathname="/jp/reservation"
   */
  it("Shallow render shoud work with pathname='/jp/reservation'", () => {
    expect(
      findByTestAttr(globalHeaderEnzymeWrapper, "global-header-wrapper")
    ).toHaveLength(1);
  });

  /**
   * Snapshot should not be updated with locale = jp
   */
  it("日本語で表示する画面が予期せず変更されていないかを確かめる", () => {
    // ご参考に：https://medium.com/7shifts-engineering-blog/testing-usecontext-react-hook-with-enzyme-shallow-da062140fc83
    // MOCK CONTEXT FOR GLOBALHEADER
    const contextValues = {
      locale: "jp",
      localeDataSource: localeDataSourceJP
    };
    jest
      .spyOn(LocaleContext, "useLocaleContext")
      .mockImplementation(() => contextValues);
    const { enzymeWrapper } = setUp();
    globalHeaderEnzymeWrapper = enzymeWrapper;
    expect(globalHeaderEnzymeWrapper).toMatchSnapshot();
  });
});
