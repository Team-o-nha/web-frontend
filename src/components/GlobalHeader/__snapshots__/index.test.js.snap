// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GlobalHeader 日本語で表示する画面が予期せず変更されていないかを確かめる 1`] = `
<Adapter
  className="global-header-wrapper clearfix"
  data-test="global-header-wrapper"
>
  <div
    className="left-menu-wrapper"
  >
    <Menu
      mode="horizontal"
      selectedKeys={
        Array [
          "Reservation",
        ]
      }
      style={
        Object {
          "lineHeight": "64px",
        }
      }
      theme="dark"
    >
      <MenuItem
        key="Home"
      >
        <Link
          to="/jp/"
        >
          ホーム
        </Link>
      </MenuItem>
      <MenuItem
        key="Reservation"
      >
        <Link
          to="/jp/reservation"
        >
          予約
        </Link>
      </MenuItem>
      <MenuItem
        key="Vacancy"
      >
        <Link
          to="/jp/vacancy"
        >
          空室問い合わせ
        </Link>
      </MenuItem>
    </Menu>
  </div>
  <div
    className="right-menu-wrapper"
  >
    <Menu
      mode="horizontal"
      selectable={false}
      style={
        Object {
          "lineHeight": "64px",
        }
      }
      theme="dark"
    >
      <MenuItem
        className="header-menu-item-right"
        key="i18next-jp"
      >
        <a
          href="/jp/reservation"
        >
          日本語
        </a>
      </MenuItem>
      <MenuItem
        className="header-menu-item-right"
        key="i18next-en"
      >
        <a
          href="/en/reservation"
        >
          English
        </a>
      </MenuItem>
    </Menu>
  </div>
</Adapter>
`;
